### PR TITLE
fix(dsa): correct packed FlashInfer top-k and backend selection semantics

### PR DESCRIPTION
**Pull request by zianglih**
## Motivation
@humansand

[PR #22851](https://github.com/sgl-project/sglang/pull/22851) introduced `--dsa-topk-backend` and integrated FlashInfer and PyTorch top-k into the fused and unfused DSA paths. This PR follows up on that integration with three focused fixes:

1. **FlashInfer packed PAGED correctness:** packed extend uses batch-global offsets for score rows while its page tables remain request-local. FlashInfer's PAGED transform applies its single `row_starts` argument to both score selection and page-table lookup, so it cannot represent this layout.
2. **Backend-selection semantics:** `SGLANG_OPT_USE_TOPK_V2` is an `sgl-kernel` implementation optimization and should not override an explicit `--dsa-topk-backend` choice. Previously, eligible decode calls could enter the SGL top-k v2 path even when another backend, such as `flashinfer` or `torch`, was selected. CUDA graph metadata construction made the same assumption when deciding whether to build the v2 plan and retain the page-size-1 table.
3. **Avoidable top-k host synchronization:** dynamic `torch.repeat_interleave` calls without `output_size` infer their output size by reading repeat counts back from CUDA. Each affected mapping is already known to produce one entry per logit row, so supplying that row count avoids the host synchronization without changing mapping semantics.

The accompanying page-size lookup change does not introduce new paging behavior. `_topk_transform_v2_paged` already consumes `DSAMetadata.real_page_table`, and both eager and CUDA graph metadata associate that table with `DSAMetadata.page_size` from the backend's `real_page_size`. Reading the table and its page size from the same metadata removes an ambient forward-context dependency while preserving the existing value.

## Modifications

- Use `sgl_kernel.fast_topk_transform_fused` as a correctness fallback for the unsupported FlashInfer PAGED case where `row_starts` is present. Unshifted PAGED and RAGGED calls continue to use FlashInfer.
- Make `SGLANG_OPT_USE_TOPK_V2` subordinate to `--dsa-topk-backend`: top-k v2 is eligible only when `sgl-kernel` is selected. Apply this policy consistently to dispatch, plan construction, and CUDA graph wide-page-table elision.
- Read the v2 transform's page size from the same `DSAMetadata` that supplies `real_page_table`, rather than consulting global forward context.
- Pass the known logit row count as `output_size` to dynamic repeat-interleave calls in FlashInfer PAGED row mapping and top-k offset construction.
- Expand backend equivalence coverage to shifted PAGED rows, including multiple query rows per request, and add a centralized regression for top-k v2 backend selection and metadata handling.

## Known Limitation

The packed PAGED fallback in this PR is numerically correct, but it is not a complete FlashInfer fused top-k path: when `row_starts` is present, the FlashInfer branch invokes `sgl_kernel.fast_topk_transform_fused`.

A fully semantically correct FlashInfer fused top-k implementation for this case requires [flashinfer-ai/flashinfer#4169](https://github.com/flashinfer-ai/flashinfer/pull/4169). That PR adds a separate `page_table_row_starts` argument so FlashInfer can use the score-window and page-table starts independently. Until SGLang adopts a FlashInfer version containing that API, `--dsa-topk-backend flashinfer` will continue to use the SGL fallback for packed PAGED rows, and the FlashInfer deterministic and tie-break settings will not apply to that shape.

The backend-selection fix in this PR is independent of that dependency: it prevents the SGL-only top-k v2 optimization from overriding an explicitly selected non-SGL backend on paths the selected backend supports.

## Accuracy Tests

Focused tests were run on an NVIDIA B200 with CUDA 13.0, FlashInfer 0.6.15.post1, and `sglang-kernel` 0.4.5:

```bash
python3 -m pytest -q --tb=short \
  test/registered/kernels/ops/attention/test_dsa_indexer.py::TestDSAIndexer::test_topk_unfused_backends_valid_selection \
  test/registered/kernels/ops/attention/test_dsa_indexer.py::TestDSAIndexer::test_topk_fused_backends_equivalence \
  test/registered/kernels/ops/attention/test_dsa_indexer.py::TestDSAIndexer::test_topk_v2_respects_topk_backend
```

Result: `3 passed, 30 subtests passed in 11.58s`.

The tests compare exact selected index sets using tie-free logits across PAGED and RAGGED layouts, shifted and unshifted rows, and multi-query batches. The multi-query fused cases also assert that every dynamic top-k repeat supplies `output_size=num_rows`, the PyTorch API contract that avoids reading `sum(repeats)` back to the host. FlashInfer tie-break modes `None`, `small`, and `large` are exercised on paths that invoke FlashInfer. Shifted PAGED cases validate the SGL fallback against the SGL backend and therefore do not exercise FlashInfer tie-breaking. No model-level accuracy benchmark was run.

`pre-commit run --all-files` also passed.

## Speed Tests and Profiling

No standalone performance benchmark was run for the repeat-interleave change. Supplying the already-known `output_size` removes PyTorch's implicit host read without changing mapping values. The packed PAGED change is a correctness fix, while the top-k v2 change restores the expected backend-selection semantics. FlashInfer remains unchanged for layouts its API can represent; only packed PAGED rows with separate score/page-table coordinate systems use the existing SGL fused transform.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). (No user-facing API change.)
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #30403592750](https://github.com/sgl-project/sglang/actions/runs/30403592750)
Latest PR Test (Extra): :x: [Run #30403592267](https://github.com/sgl-project/sglang/actions/runs/30403592267)

### CODE REVIEW

**Review by Fridge003 — APPROVED** [review]

**Comment by Fridge003** [comment]
/rerun-test test/registered/models_e2e/test_dsa_glm52_tp_mtp.py test/registered/models_e2e/test_dsa_glm52_dp_mtp.py test/registered/models_e2e/test_dsa_glm52_nvfp4_dp_mtp.py test/registered/models_e2e/test_dsa_glm52_nvfp4_tp_mtp.py  test/registered/cp/test_glm52_cp_index_share.py test/registered/models_e2e/test_dsa_glm52_hisparse.py test/registered/models_e2e/test_dsa_glm52_cache_layer_split.py test/registered/kernels/ops/attention/test_dsa_indexer.py
